thewalrus/tests/test_fock_gradients.py

```diff
@@ -148,25 +124,122 @@ def test_Kgate():
+    assert np.allclose(T, expected, atol=tol, rtol=0)
+
+
+def test_Sgate_values_complex(tol):
+    """Tests the correct construction of the single mode squeezing operation"""
+    r = 0.5
+    theta = 0.7
+    cutoff = 5
+    # This data is obtained by using qutip
+    # np.array(squeeze(40,0.5*np.exp(1j*0.7)).data.todense())[0:5,0:5]
+    expected = np.array(
+        [
+            [0.94171062 + 0.0j, 0.0 + 0.0j, 0.23535661 - 0.19823814j, 0.0 + 0.0j, 0.02093159 - 0.12135894j,],
+            [0.0 + 0.0j, 0.83512676 + 0.0j, 0.0 + 0.0j, 0.36151137 - 0.30449682j, 0.0 + 0.0j],
+            [-0.23535661 - 0.19823814j, 0.0 + 0.0j, 0.64005396 + 0.0j, 0.0 + 0.0j, 0.42261153 - 0.35596078j,],
+            [0.0 + 0.0j, -0.36151137 - 0.30449682j, 0.0 + 0.0j, 0.38926873 + 0.0j, 0.0 + 0.0j],
+            [0.02093159 + 0.12135894j, 0.0 + 0.0j, -0.42261153 - 0.35596078j, 0.0 + 0.0j, 0.12407853 + 0.0j,],
+        ]
+    )
+    T = Sgate(r, theta, cutoff)[0]
+    assert np.allclose(T, expected, atol=tol, rtol=0)
+
+
+def test_BS_selection_rules(tol):
+    r"""Test the selection rules of a beamsplitter.
+    If one writes the beamsplitter gate of :math:`U` and its matrix elements as
+    :math:`\langle m, n |U|k,l \rangle` then these elements
+    are nonzero if and only if :math:`m+n = k+l`. This test checks
+    that this selection rule holds.
+    """
+    cutoff = 4
+    T = BSgate(np.random.rand(), np.random.rand(), cutoff)[0]
+    for m in range(cutoff):
+        for n in range(cutoff):
+            for k in range(cutoff):
+                for l in range(cutoff):
+                    if m + n != k + l:  # Check that there are the same total number of photons in the bra and the ket
+                        assert np.allclose(T[m, n, k, l], 0.0, atol=tol, rtol=0)
+
+
+def test_BS_hong_ou_mandel_interference(tol):
+    r"""Tests Hong-Ou-Mandel interference for a 50:50 beamsplitter.
+    If one writes :math:`U` for the Fock representation of a 50-50 beamsplitter
+    then it must hold that :math:`\langle 1,1|U|1,1 \rangle = 0`.
+    """
+    cutoff = 2
+    phi = 2 * np.pi * np.random.rand()
+    T = BSgate(np.pi / 4, phi, cutoff)[0]  # a 50-50 beamsplitter with phase phi
+    assert np.allclose(T[1, 1, 1, 1], 0.0, atol=tol, rtol=0)
+
+
+def test_S2_selection_rules(tol):
+    r"""Tests the selection rules of a two mode squeezing operation.
+    If one writes the squeezing gate as :math:`S_2` and its matrix elements as
+    :math:`\langle p_0 p_1|S_2|q_0 q_1 \rangle` then these elements are nonzero
+    if and only if :math:`p_0 - q_0 = p_1 - q_1`. This test checks that this
+    selection rule holds.
+    """
+    cutoff = 5
+    s = np.arcsinh(1.0)
+    phi = np.pi / 6
+    T = S2gate(s, phi, cutoff)[0]
+    for m in range(cutoff):
+        for n in range(cutoff):
+            for k in range(cutoff):
+                for l in range(cutoff):
+                    if m - n != k - l:
+                        assert np.allclose(T[m, n, k, l], 0, atol=tol, rtol=0)
```

Comment:
same comment here re: vectorization and removing the loops

Reply:
ditto!